### PR TITLE
Feature 82. Create shared app components. narek

### PR DESCRIPTION
**Pull request by narekacy**
Developers can use shared components to create their own view easy and flexible. Furthermore, using 'ShowMenu' component will help to show 'Restaurants' and 'Dishes' search in simple way

### CODE REVIEW

**Comment by VanderbiGit** [comment]
You not shared map and restaurant.

File: client/src/app.js [hunk 1]
```diff
@@ -1,49 +1,28 @@
...
-import About from './pages/about';
-import Contact from './pages/contact';
-import ListRestaurant from './pages/restaurant-list';
-import Login from './pages/login';
-import Registartion from './pages/registartion';
-import Map from './pages/map';
 import NoMatch from './pages/no-match';
-import Events from './pages/events';
-import Restaurant from './components/restaurant-item';
+import info from './components/info/rotes';
```
Comment (vitaliimak) [inline client/src/app.js]: You use this array only in one place. Why you send it in separate file?

File: client/src/components/info/rotes.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
```
Comment (vitaliimak) [inline client/src/components/info/rotes.js]: You should be pay attention in file naming. I hope we don't need "rote" in our project

File: client/src/app.js (hunk 1 above)
Comment (vitaliimak) [reply client/src/app.js]: It is very inconveniently to go to another file when I only need to add one line in `app.js`

File: client/src/components/shared/button/reset.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import React, { Component } from 'react';
+import SearchMenuButton from './search-menu-button';
+
+class Reset extends Component {
```
Comment (vitaliimak) [inline client/src/components/shared/button/reset.js:4]: Don't you think that it is to difficult? Create shared components for so small piece of code? There will be buttons save, delete, update, search, and many other.

File: client/src/components/shared/restaurant/restaurant-item.js [hunk 4]
```diff
@@ -1,9 +1,9 @@
...
 import { Tab, Tabs, Container } from 'react-bootstrap';
 import PropTypes from 'prop-types';
-import Api from '../services/api';
+import Api from '../../../services/api';
 import About from './restaurant-about';
-import '../style/restaurant-item.css';
+import '../../../style/restaurant-item.css';
 
 class Restaurant extends Component {
   constructor(props) {
```
Comment (vitaliimak) [inline client/src/components/shared/restaurant/restaurant-item.js]: I don't understand what happens and why
image

**Review by vitaliimak — CHANGES_REQUESTED** [review]
As you send a lot of constants in separate files (I think it is bad idea if we use this constants only in one places each) you shouldn't place those files in component folder. You can create `/client/src/shared/constants` folder, or something similar. It makes your code structure simplier

I think this code brings more difficulties in our project. We have very few entities (`Restautant and Events`) but we will get very huge and difficulty structure.
```
client/src
├── app.js
├── app.test.js
├── components
│   ├── info
│   │   ├── events.js
│   │   ├── navigation.js
│   │   └── rotes.js
│   ├── openlayers-map.js
│   └── shared
│       ├── button
│       │   ├── date-picker.js
│       │   ├── reset.js
│       │   └── search-menu-button.js
│       ├── dropdown
│       │   ├── context.js
│       │   ├── date-menu.js
│       │   └── view.js
│       ├── event
│       │   ├── card-results.js
│       │   └── card-view.js
│       ├── navigation
│       │   ├── navigation-bar.js
│       │   ├── navigation-item.js
│       │   └── navigation-route.js
│       ├── restaurant
│       │   ├── restaurant-about.js
│       │   └── restaurant-item.js
│       └── search
│           ├── filter.js
│           ├── header.js
│           ├── input.js
│           ├── search-menu.js
│           └── title.js
├── index.js
├── pages
│   ├── about.js
│   ├── contact.js
│   ├── events.js
│   ├── home.js
│   ├── login.js
│   ├── map.js
│   ├── no-match.js
│   ├── registration.js
│   └── restaurant-list.js
├── services
│   └── api.js
├── service-worker.js
├── setup-tests.js
└── style
    ├── app.css
    ├── dropdown.css
    ├── events-page.css
    ├── m-button.css
    ├── openlayers-styles.css
    ├── restaurant-about.css
    └── restaurant-item.css
```
I think this code won't make my developer experience simpler and we should re-think about this task

**Comment by narekacy** [comment]
> You not shared map and restaurant.

Restaurant is already sharing in `components/shared/restaurant`

As for map, I've moved it from `components/` to `components/shared/` already, thanks.

File: client/src/app.js (hunk 1 above)
Comment (narekacy) [reply client/src/app.js]: > You use this array only in one place. Why you send it in separate file?

We should separate 'view-info' and component. This is temporarily decision. The previos one is not the best that should be in React.js component.

Comment (narekacy) [reply client/src/app.js]: > It is very inconveniently to go to another file when I only need to add one line in `app.js`

This is not one line, `info` is app's routes representation here.

File: client/src/components/info/rotes.js (hunk 2 above)
Comment (narekacy) [reply client/src/components/info/rotes.js]: You have already used 'route' in **[Profile](https://github.com/Social-projects-Rivne/Lunch_Time/blob/0bfb700b2274cb15a0846a0c6ec42888bcd8e3c2/client/src/components/profile.js#L83-L93)** component. As for file name, what we can name it, for example? I'll change it, just give an example.

Comment (vitaliimak) [reply client/src/components/info/rotes.js]: Route and rote are different ;)

File: client/src/components/shared/button/reset.js (hunk 3 above)
Comment (narekacy) [reply client/src/components/shared/button/reset.js:4]: This is right decision due to React.js components in application. If we use our 'button' with same style but different name 100 times, we will write redundand code lines for their creation. Moreover, we discussed this with our team and decided to use 'Button' shared component.

File: client/src/components/data/navigation-bar.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  { link: '/restaurants', name: 'ListRestaurant' },
+  { link: '/events', name: 'Events' },
+  { link: '/map', name: 'Map' },
+  { link: '/about', name: 'About' },
+  { link: '/contact', name: 'Contact' },
+  { link: '/login', name: 'Login' },
+  { link: '/registartion', name: 'Registartion' },
+];
+
+const navigationBar = data.map((e) => (
```
Comment (VanderbiGit) [inline client/src/components/data/navigation-bar.js]: You create navigation Bar in cycle. Why you did not this in restaurant. Can you commit new changes.

File: client/src/components/shared/restaurant/restaurant-item.js (hunk 4 above)
Comment (narekacy) [reply client/src/components/shared/restaurant/restaurant-item.js]: > As you send a lot of constants in separate files (I think it is bad idea if we use this constants only in one places each) you shouldn't place those files in component folder. You can create `/client/src/shared/constants` folder, or something similar. It makes your code structure simplier
> 
> I think this code brings more difficulties in our project. We have very few entities (`Restautant and Events`) but we will get very huge and difficulty structure.
> 
> ```
> client/src
> ├── app.js
> ├── app.test.js
> ├── components
> │   ├── info
> │   │   ├── events.js
> │   │   ├── navigation.js
> │   │   └── rotes.js
> │   ├── openlayers-map.js
> │   └── shared
> │       ├── button
> │       │   ├── date-picker.js
> │       │   ├── reset.js
> │       │   └── search-menu-button.js
> │       ├── dropdown
> │       │   ├── context.js
> │       │   ├── date-menu.js
> │       │   └── view.js
> │       ├── event
> │       │   ├── card-results.js
> │       │   └── card-view.js
> │       ├── navigation
> │       │   ├── navigation-bar.js
> │       │   ├── navigation-item.js
> │       │   └── navigation-route.js
> │       ├── restaurant
> │       │   ├── restaurant-about.js
> │       │   └── restaurant-item.js
> │       └── search
> │           ├── filter.js
> │           ├── header.js
> │           ├── input.js
> │           ├── search-menu.js
> │           └── title.js
> ├── index.js
> ├── pages
> │   ├── about.js
> │   ├── contact.js
> │   ├── events.js
> │   ├── home.js
> │   ├── login.js
> │   ├── map.js
> │   ├── no-match.js
> │   ├── registration.js
> │   └── restaurant-list.js
> ├── services
> │   └── api.js
> ├── service-worker.js
> ├── setup-tests.js
> └── style
>     ├── app.css
>     ├── dropdown.css
>     ├── events-page.css
>     ├── m-button.css
>     ├── openlayers-styles.css
>     ├── restaurant-about.css
>     └── restaurant-item.css
> ```
> 
> I think this code won't make my developer experience simpler and we should re-think about this task

Check Step 1 **[here](https://reactjs.org/docs/thinking-in-react.html)** and **[this article](https://medium.com/@level_out/everything-is-a-component-cf9f469ad981)**, please.

File: client/src/components/data/navigation-route.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import Registartion from '../../pages/registartion';
+import Map from '../../pages/map';
+import Contact from '../../pages/contact';
+import Events from '../../pages/events';
+
+const data = [
+  { path: '/about', component: <About /> },
+  { path: '/events', component: <Events /> },
+  //  { path: '/restaurants', component: <Restaurant match="id" /> },
+  //  { path: '/restaurants', component: <ListRestaurant  match={}/> },
```
Comment (VanderbiGit) [inline client/src/components/data/navigation-route.js]: You need this?

File: client/src/components/info/rotes.js (hunk 2 above)
Comment (narekacy) [reply client/src/components/info/rotes.js]: This is for my armenian accent, sorry. Already fixed :)

File: client/src/components/shared/search-button.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import Button from 'react-bootstrap/Button';
```
Comment (VanderbiGit) [inline client/src/components/shared/search-button.js]: Use { Button }
https://codeburst.io/es6-destructuring-the-complete-guide-7f842d08b98f

File: client/src/components/dropdown-group.js [hunk 8]
```diff
@@ -28,7 +28,10 @@ class DropdownGroup extends React.Component {
         <DropdownWithDate />
 
         {/* Reset filters */}
```
Comment (VanderbiGit) [inline client/src/components/dropdown-group.js]: You need comment in render?

File: client/src/components/shared/event-filter-bar.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import SearchFilterToolbar from './search-filter-toolbar';
+import eventData from '../data/event-data';
+
+class EventFilterBar extends React.Component {
+  render() {
+    // const bool {is date} ?
```
Comment (VanderbiGit) [inline client/src/components/shared/event-filter-bar.js]: ?

File: client/src/components/shared/dropdown/context.js [hunk 10]
```diff
@@ -4,9 +4,9 @@ import View from './view';
 
 class Context extends Component {
   render() {
-    const { data } = this.props;
+    const { info } = this.props;
     return (
-      data && data.length > 0 && data.map((e) => {
+      info && info.length > 0 && info.map((e) => {
```
Comment (VanderbiGit) [inline client/src/components/shared/dropdown/context.js]: if(info.length > 0) {.... code}
it's more shortly

**Review by VanderbiGit — COMMENTED** [review]
Delete not necessary commentary in code.

File: client/src/components/data/navigation-route.js (hunk 6 above)
Comment (narekacy) [reply client/src/components/data/navigation-route.js]: Check my last commit.This one is test/temporary by mistake.

File: client/src/components/data/navigation-bar.js (hunk 5 above)
Comment (narekacy) [reply client/src/components/data/navigation-bar.js]: Check my last commit. This one is deprecated.

**Comment by VanderbiGit** [comment]
In component Map you hardcode coordinate city Rivne. More better you can put coordinate in props.

File: client/src/components/shared/search-button.js (hunk 7 above)
Comment (narekacy) [reply client/src/components/shared/search-button.js]: Bro, you advise is just in two words, but you give a huge article to read. Just give the quote and everything will be clear. I can't see why we need to use { Button } but not Button here.

File: client/src/components/dropdown-group.js (hunk 8 above)
Comment (narekacy) [reply client/src/components/dropdown-group.js]: I guess this part of code was written by another developer. My task is to create shared components due to that ones we can reuse them. By the way, most part of our app components need to be rewritten, but this is not my current task.

File: client/src/components/shared/event-filter-bar.js (hunk 9 above)
Comment (narekacy) [reply client/src/components/shared/event-filter-bar.js]: What?

File: client/src/components/shared/search-button.js (hunk 7 above)
Comment (VanderbiGit) [reply client/src/components/shared/search-button.js]: it's a new feature in ES6 and more shortly

**Comment by narekacy** [comment]
> In component Map you hardcode coordinate city Rivne. More better you can put coordinate in props.

This 'Map' component was created for our MVP. To be perfect, we need will have an array of props or something else. We even don't need any default position right now. What you say, it is for another task.

But I will do that.

File: client/src/components/shared/dropdown/context.js (hunk 10 above)
Comment (narekacy) [reply client/src/components/shared/dropdown/context.js]: Done!

**Review by narekacy — COMMENTED** [review]
> You not shared map and restaurant.

Done.

File: client/src/components/shared/search-button.js (hunk 7 above)
Comment (narekacy) [reply client/src/components/shared/search-button.js]: Done.

**Review by Andrii35 — COMMENTED** [review]
I have wrote this in the end of line. 
"This branch has conflicts that must be resolved
 Conflicting files
client/src/app.js
client/src/components/navigation-bar.js"
By other pull requests i don't see something like that.

**Comment by narekacy** [comment]
> I have wrote this in the end of line.
> "This branch has conflicts that must be resolved
> Conflicting files
> client/src/app.js
> client/src/components/navigation-bar.js"
> By other pull requests i don't see something like that.

This is about another merging into dev after my pull request. I will update in soon.

File: client/src/components/shared/dropdown/context.js (hunk 10 above)
Comment (vitaliimak) [reply client/src/components/shared/dropdown/context.js]: It is JavaScript. You can even use.
```
if (data.length) {
  // code
}
```
But you should certain that `data` is Array. If `data === null or data === undefined or data is Object` you'll get an error

File: client/src/components/dropdown-group.js (hunk 8 above)
Comment (vitaliimak) [reply client/src/components/dropdown-group.js]: If you see so small mistake you can fix this mistake. It will make codebase cleaner

Comment (narekacy) [reply client/src/components/dropdown-group.js]: Ok, got it.

**Comment by narekacy** [comment]
Bux are already fixed. Please, check my pull request again.

File: client/src/components/shared/search/filter.js [hunk 11]
```diff
@@ -7,11 +7,11 @@ import ResetButton from '../button/reset';
 
 class Filter extends Component {
   render() {
-    const { info } = this.props;
+    const { info, showDate } = this.props;
     return (
       <ButtonToolbar className="justify-content-center">
         <Context info={info} />
-        <DateMenu />
+        {showDate === true && <DateMenu /> }
```
Comment (VanderbiGit) [inline client/src/components/shared/search/filter.js]: You can write {showDate && DateMenu }
its shortly

Comment (narekacy) [reply client/src/components/shared/search/filter.js]: Thank you. Fixed!

File: client/src/components/shared/navigation-item.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import React, { Component } from 'react';
+import { Nav } from 'react-bootstrap';
+import { Link } from 'react-router-dom';
+import PropTypes from 'prop-types';
+
+class NavigationItem extends Component {
+  render() {
+    const { clName, link, name } = this.props;
+    return (
+      <nav>
```
Comment (VanderbiGit) [inline client/src/components/shared/navigation-item.js]: You not need nav

Comment (narekacy) [reply client/src/components/shared/navigation-item.js]: Thanks, fixed!

File: client/src/components/shared/search/header.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+class Header extends Component {
+  render() {
+    const {
+      containerClassName, h1ClassName, title, placeHolder,
+    } = this.props;
+    return (
+      <Container fluid className={containerClassName}>
+        <h1 className={h1ClassName}>
+          <Title title={title} />
+          <Input placeHolder={placeHolder} />
```
Comment (VanderbiGit) [inline client/src/components/shared/search/header.js]: h1 it's only Headline

Comment (narekacy) [reply client/src/components/shared/search/header.js]: Thanks, fixed!

**Review by Andrii35 — DISMISSED** [review]
Good!

**Review by VanderbiGit — DISMISSED** [review]

**Review by Melnyk9506 — DISMISSED** [review]

File: server/mvnw [hunk 14]
```diff
@@ -201,7 +201,7 @@ fi
 
 ##########################################################################################
 # Extension to allow automatically downloading the maven-wrapper.jar from Maven-central
-# This allows using the maven wrapper in projects that prohibit checking in binary data.
+# This allows using the maven wrapper in projects that prohibit checking in binary info.
```
Comment (vitaliimak) [inline server/mvnw:204]: You should be attentive when you automatically replace words

File: client/src/components/info/events.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
+const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July',
```
Comment (vitaliimak) [inline client/src/components/info/events.js:1]: Why those constant are placed in info folder? Why this folder is placed in components? How developers can connect words components, info and React component Events?

File: client/src/components/event/results.js [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+              <CardView key={event.id} event={event} isFetching={isFetching} />
+            ))}
+          </CardDeck>
+        ) : (
+          <Container className={spinnerContainerClassName}>
+            <ButtonToolbar className={buttonToolbarClassName}>
+              <Spinner animation={spinnerAnimation} variant={spinnerVariant} />
+            </ButtonToolbar>
+          </Container>
+        )}
```
Comment (vitaliimak) [inline client/src/components/event/results.js:30]: This ternar operator is looked a little big. https://github.com/airbnb/javascript#comparison--nested-ternaries

File: client/src/components/event/filter-bar.js [hunk 17]
```diff
@@ -50,8 +47,8 @@ class DropdownGroup extends React.Component {
   }
```
Comment (vitaliimak) [inline client/src/components/event/filter-bar.js:47]: Shared button?

**Review by Andrii35 — DISMISSED** [review]

File: client/src/components/event/filter-bar.js [hunk 18]
```diff
@@ -50,8 +47,8 @@ class DropdownGroup extends React.Component {
...
 }
```
Comment (vitaliimak) [inline client/src/components/event/filter-bar.js:48]: @yevchuk `window.location.reload()` reloads all page it isn't SPA approach. I've created task https://trello.com/c/uF3svZRL/96-91-fix-button-reset-in-events

**Review by vitaliimak — COMMENTED** [review]
Too difficult for me, I can't approve :(

File: client/src/components/event/filter-bar.js (hunk 17 above)
Comment (vitaliimak) [reply client/src/components/event/filter-bar.js:47]: According to new task description you don't need to do refactoring, so my previous comment isn't relevant

File: client/src/components/info/events.js (hunk 15 above)
Comment (vitaliimak) [reply client/src/components/info/events.js:1]: As those constants aren't React Components you shouldn't change previous behavior (refactoring). Or if you absolutely sure that those constants are used in different places (right now, not in future) you can move those in separate files in folders like `shaded/constants`.
If you sure that those constants are used only in one component (Search component) you shouldn't move them from this component.

The same message concerns other constants, which are used only in one places, but you moved those in separate files

**Comment by Melnyk9506** [comment]
I think in the future I think in the future it will help to improve design changes and not only

**Comment by VozNick** [comment]
I think it would be better to make constants not shared, because while we develop some tasks, we will edit those files many times. And in feature we can refactor code to make it look more better.

File: client/src/components/info/events.js (hunk 15 above)
Comment (narekacy) [reply client/src/components/info/events.js:1]: On front part of our app, we show these categories as property values of our entities from back part (moths, event categories, etc). So, I guess we should get it from back part. BUT: due to my poor experience in this question, I just wanted to move these constants up. 

I agree that this in not best decision, but I moved 'hard-code' up, and, finally, some day [may be, I hope] we will get it from back. I want to believe that our code should be less changable and independent. 

Due to our requirements we will need this values in profile history (i.e. search my events). BUT: again, we will get this from back part also. So, if we had no requirements for it, I would probably do the same.

That's why our components shoud render with same code even with different incoming data. It means that we should move 'modifile' level up and up, not inside developer. I like your logic and opinion. I just wanted to show that component should be fully independent.

And yes, we can move it to folder up and name it constants, it is good idea.

**Review by vitaliimak — DISMISSED** [review]

**Review by VozNick — DISMISSED** [review]

**Comment by r-yevchuk** [comment]
Scrollbar is broken on the events page. 
And would you remove this semicolon.

**Comment by narekacy** [comment]
> Scrollbar is broken on the events page.
> And would you remove this semicolon.
> 

Fixed

**Comment by r-yevchuk** [comment]
> > Scrollbar is broken on the events page.
> > And would you remove this semicolon.
> > 
> 
> Fixed

Fine. But scrolling still doesn't work

**Review by r-yevchuk — APPROVED** [review]

**Review by VozNick — APPROVED** [review]

**Review by Melnyk9506 — APPROVED** [review]

**Pull request merged**